### ISSUE TITLE
Improve the design of stacks, state and projects

### ISSUE THREAD

**Issue by atrauzzi** [pulumi/pulumi#8402]
- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently, Pulumi maintains a 1:1 relationship between stacks and projects.  This means that if I have multiple projects targeting the same "environment", instead of being able to define one stack and one corresponding configuration file; I'm on the hook for one stack per project, per environment.

> note: An _environment_ in this case is like a logical handle for a grouping of resources intended to collaborate together.  Frequently represented by a Google Cloud project or Azure resource group.

So this means if I - rather conservatively - have the following projects:

 - Core
 - API
 - Ingress

Then I end up having to make three `Pulumi.{environment}-{project}.yaml` files in each respective project folder for just one environment:

 - `Pulumi.dev-core.yaml`
 - `Pulumi.dev-api.yaml`
 - `Pulumi.dev-ingress.yaml`

Multiply this by - let's assume - an average of three environments per overall architecture:

 - `Pulumi.dev-core.yaml`
 - `Pulumi.dev-api.yaml`
 - `Pulumi.dev-ingress.yaml`
 - `Pulumi.test-core.yaml`
 - `Pulumi.test-api.yaml`
 - `Pulumi.test-ingress.yaml`
 - `Pulumi.prod-core.yaml`
 - `Pulumi.prod-api.yaml`
 - `Pulumi.prod-ingress.yaml`

...suddenly I have a proliferation of `yaml` and countless opportunities for data to duplicate and require painful amounts of manual management.  Keep in mind, I'm only using an example that consists of three projects.  If I use my own current setup, I have *six* projects.  Using even just *two* environments that's *twelve* yaml files. :dizzy: 

#### The Problem

I think the issue is that the Pulumi state database seems to equate stacks with projects.  Whereas, it might be necessary to introduce some notion of an "environment" at the state store level.  Right now, Pulumi has you "writing a stack and then creating a stack", which is confusing.  The first instance of _stack_ [being the code](https://github.com/pulumi/pulumi/issues/8371) and the second instance of stack being the `yaml` configuration file.  For lack of a better way of expressing it, this seems like a structural flaw in how Pulumi organizes its own data, which causes some dissonance.

If we use our imagination a little, introducing terminology to represent _environments_ changes the paradigm a little.  You wouldn't "create a stack and then create a stack".  You'd "author a stack and instantiate it in an environment".  Then - by extension - you'd be able to "...instantiate multiple stacks in an environment".

#### My Current Workaround

I've been able to clumsily simulate this functionality taking three steps:

 - Writing a bash script that does a little brute force management of stacks and calling them
 - Creating a *single* `Pulumi.{environment}.yaml` file which the script uses
   - This `yaml` file uses multiple `config:` sections
 - Specifying config section when instantiating `Pulumi.Config` so that I can reuse configuration between projects/stacks (there's that dissonance again!)

Here's my terrible, yet extremely effective bash script:

```bash
#!/bin/bash

projectCamel=${1}
projectKebab="$(echo "${projectCamel}" | sed 's/\(.\)\([A-Z]\)/\1-\2/g' | tr '[:upper:]' '[:lower:]')"
environment=${2}
projectStudly="${projectCamel^}"

# note: I'd like to be able to uncomment this line and comment out the next one and have Pulumi still work.
#stackName="${environment}"
stackName="${environment}-${projectKebab}"

cd ./*Deployment."${projectStudly}" || exit

{
    pulumi stack init "${stackName}"
} &> /dev/null

{
    rm Pulumi."${stackName}".yaml
    ln -s "../Pulumi.${environment}.yaml" "./Pulumi.${stackName}.yaml"
} &> /dev/null

pulumi up \
    --emoji \
    --stack="${stackName}"
    #    --non-interactive --yes \

rm "Pulumi.${stackName}.yaml"
```

Overall?  The number of lines of config I've spared myself with this script alone prove that there's an opportunity to greatly improve the experience of config management and structure in Pulumi-based source repositories.

Hopefully I've used all the right terminology, but feel free to ask questions to clarify or offer corrections. :smile:

**Comment by Frassle** [comment]
I think https://github.com/pulumi/pulumi/issues/2307 would maybe go someway towards helping with this. The environment idea is a bit more complicated as that sounds like values for some stacks across different projects so it's not purely hierarchical.

**Comment by atrauzzi** [comment]
It's less about _values from stacks_ and more about two things:

 - Values being available across stacks (IMO, config values shouldn't conceptually belong to any one stack)
 - Stacks being able to support multiple projects
 
In theory, the environment idea was just a nice way to group that concept into something that could be logically expressed.

I have no :microscope: _science_ :test_tube: for this, but I think what I'm describing is a mental model that most people probably approach Pulumi expecting.

**Comment by lukehoban** [comment]
> Currently, Pulumi maintains a 1:1 relationship between stacks and projects

Just a note that this isn't quite true.  Pulumi projects are 1:N with stacks - each project can have N stacks associated with it.  Each `Pulumi.yaml` file represents the root of a project, and next to that there can be N `Pulumi.<stackname>.yaml` stacks associated with that project.

> think the issue is that the Pulumi state database seems to equate stacks with projects. 

Are you referring here to the design of the self-managed backends, and specifically the issues with those backends called out in https://github.com/pulumi/pulumi/issues/2522 and https://github.com/pulumi/pulumi/issues/2814?  Or are you referring to something deeper in the core Pulumi model?

> suddenly I have a proliferation of yaml and countless opportunities for data to duplicate and require painful amounts of manual management

In that case of 12 files, in principle you might legitimately want configuration to differ across each of those 12 stacks, right?  But you want a way to pull defaults up into the project?  And/or into an orthogonal notion of "environment" that you could associate with a stack?  The former (pulling configuration into a project, or even up into an organization) is what https://github.com/pulumi/pulumi/issues/2307 is tracking as @Frassle notes.  The latter is something that hasn't come up quite as often, but I think is a very valid scenario - which sounds related to the ideas explored in https://github.com/pulumi/pulumi/issues/7048.

**Comment by atrauzzi** [comment]
No problem :slightly_smiling_face: 

> ...Pulumi projects are 1:N with stacks - each project can have N stacks associated with it.

This demonstrates some of the problem I highlight in #8371.  The word _stack_ is ambiguous and has two meanings in the Pulumi glossary.  One is an instantiation of a project, which then seems to become fused to _stack state_.  But then there's the _code_ definition of _stack_ as in `public class MyStack extends Stack`.

What I'm referring to here is the _stack state_.  And _stack state_ can only have one project.  This is evidenced by the fact that if I configure two projects to use the same stack name and try to do `pulumi up` both, the second one will try and delete everything from the first.
Maybe #2522 and #2814 are to blame, but I'm not familiar with the differences as I've only ever used the local store. If that is the case, the first ticket has been open since 2019, I hope this is being addressed with a patch release at this point. :worried: 

> In that case of 12 files, in principle you might legitimately want configuration to differ across each of those 12 stacks, right?

I am more likely to have configuration that I wish to reuse between them and I'm less likely to want to manage one-file-per-project-per-environment.  I'd much rather maintain *one* `yaml` file, manage all my configuration there and then instruct projects to read from only the sections they are concerned with.

But I think my preference there might be better undestood with this next section...

> The latter is something that hasn't come up quite as often, but I think is a very valid scenario - which sounds related to the ideas explored in #7048.

To be honest, I think the filesystem needs to be removed from the equation as it causes a lot of confusion.  Any call to pulumi should just be fed configuration, and where that configuration originates from shouldn't influence its internal data model.  Right now, a stack name *couples a project, a config file and state*.

---

Right now I also seem to be struggling with doing `pulumi up` for the same stack twice. My script above symlinks my central configuration to a uniquely named stack in the project directory.  But for some reason, I'm still required to do `pulumi init` and/or `pulumi init` can't just initialize the state store without attempting to write a `Pulumi.*.yaml` to the filesystem; Which I've already done for it.

The coupling between the filesystem and the state store as well as the limitation of one project per _state stack_ seems to bake in assumptions that probably don't need to exist.

**Comment by lukehoban** [comment]
> if I configure two projects to use the same stack name and try to pulumi up both one after the other, the second one will try and delete everything from the first.
Maybe #2522 and #2814 are to blame, but I'm not familiar with the differences as I've only ever used the local store.

Yes - the issue you are describing these is exactly due to those issues (#2522 specifically).  I have a feeling that the majority of the issue you are experiencing here is due to that.  For the service backend, you do not need to include the project name in the stack name - the project is a "namespace" for stacks.

Our goal, as #2522 notes, is to align the filestate backends with the behaviour of the httpstate backend (the one the Pulumi Service implements).  That will provide a single shared backend store across projects, with project-baed namespacing of stack identity.  I am fairly sure this would address almost all of the points you have raised here (with the exception of hierarchical configuration as a way to avoid some configuration copy-paste, but it sounds like this is not as significant of a concern for you).

I expect this issue should likely be resolved as a duplicate of #2522, but given there's a lot of great details here, will leave this open for now to capture any further conversation.

**Comment by atrauzzi** [comment]
There's a little bit pertaining to how the CLI behaves regarding stack files that I'm still not sure about. Like how Pulumi won't show me my stacks unless I'm in a project. Even then, it doesn't matter which project I'm in.

Then also how Pulumi requires me to `init` a stack, even though I've already populated a configuration.

There's definitely something to how Pulumi over-relies on files in `pwd` which could be backed off and night help people understand how they can work with stacks in state.

> the project is a "namespace" for stacks...

This part I'm not so sure about. It looks like Pulumi only has one code stack per project. The combination of which becomes one project per **state stack**.

In my C# project, I define multiple `.csproj`, each with their own `Pulumi.yaml`. But each one of those only ever ends up instantiating one `: Stack` (**code stack**).

**Comment by atrauzzi** [comment]
@lukehoban - On a whim, I switched over to trying the Pulumi state backend and still encountered the requirement to name stacks differently between projects.

I really think there's something seriously broken in how Pulumi organizes state internally in combination with some odd or poorly chosen CLI behaviours.

Would you or any other Pulumi stakeholders be interested in a screenshare with me to walk through things?  I'm happy to have a deeper conversation on this with examples we can pick at...

**Comment by chrisui** [comment]
Looking for a solution to my issues with stacks between local and cloud backends I stumbled on this issue which I think is related (at least in the sentiment of "there seems to be something not quite right with how stacks/projects/orgs are designed".

[See my unanswered question in slack](https://pulumi-community.slack.com/archives/C84L4E3N1/p1637855798008300).

> How are people, if at all, managing the stack name/referencing variation between local and default state backends? On the cloud a stack reference is derived from project name and stack name (ie. [ORG/]PROJECT/STACK ) but locally it’s just the stack (ie. STACK). This means I end up with conflicts because every project has a "local" stack. You can specify a location per login but I presume logging in separately for every project is going to be a pain and mean I cannot do anything in parallel (a non-starter).
Is my only solution to have an inconsistent stack name of “local-PROJECT” (Pulumi.production.yml next to Pulumi.local-projec.yml ) or am I missing something else? Slightly unfortunate that I also have to propagate logic for handling name convention difference into my runtime for StackReference too.

**Comment by ghostsquad** [comment]
I would like to chime in on this, as I immediately ran into ALL of the same issues described in this thread, which for clarity, I'll briefly mention:

- `pulumi login` & `pulumi stack select` - these are stateful operations - parallelization is impossible - this is the same problem of using `kubectl` without explicitly mentioning the context (i.e. `kubectl --context my-cluster`), or the aws cli, without passing in `aws --profile` or setting the `AWS_PROFILE` env var explicitly for that session. But it's worse than that... 
- `pulumi login s3://pulumi-myproject` describes where the state for the project is stored/managed. Why does this value (`s3://pulumi-myproject`) show up in `$HOME/.pulumi` and not in the project itself? How do I share this with other developers, or via the CI system? In terraform, it's defined in the `terraform {}` block, and checked in with everything else.
- In regards to the stack naming - I assumed, (and I'm sure I'm still right), that I want different "backends" for different projects, i.e. `s3://pulumi-app1` and `s3://pulumi-app2` - this is important for "least privilege security". If you don't need to know about `app1`, why would those states be colocated? This is also similar to how terraform works. Ok, so separate backends does allow me to have identically named "stacks" (e.g. dev, staging, prod) between each application. However, it's still a pain to manage the `pulumi login` and `pulumi stack select` process. I want something like terraform, or terragrunt.

```
    ├── pulumi.yml <--- describes project config, such as `s3://pulumi-myapp`
    ├── prod
    │   └── pulumi.yml <--- describes the stack config, i.e. `pulumi stack select prod`
    ├── qa
    │   └── pulumi.yml
    └── stage
        └── pulumi.yml
```

The expectation would be running `pulumi up` from the `prod` folder, just does the right thing, and I have no need explicitly login or change stacks or anything like that.

Even with an incremental move to this model, the request would be to include something some flags to `pulumi up` such as:

```
--backend-url
```

and

```
--stack
```

Then I can write a wrapper _at least_ that is _stateless_.

**Comment by ghostsquad** [comment]
> There's definitely something to how Pulumi over-relies on files in `pwd` which could be backed off and might help people understand how they can work with stacks in state.

I don't think there's necessarily anything particularly wrong about relying on files in `pwd`, after all `terraform` is the same way. There _is_ room though to be more like `kubectl` in some ways. I'm imagining a scenario like this

`pulumi status --backend s3://pulumi-myapp --stack dev` and pulumi can go look at the stack state, and the _real_ state, and tell me if there's any drift between the IRL state of resources, and what's described in the pulumi state.

This process is similar to `kubectl diff -f myapp.yaml` - In that it doesn't matter where you are running the command from (so your PWD doesn't matter), but you _can_ tell kubectl what to compare to. Similarly, `kubectl diff -f .` _does_ care about your current directory, but it's very obvious of that from the command.

**Comment by prom3theu5** [comment]
I actually put something together for a basic hierarchy with Stack config.
Forgive me if there is a better way - nly been using pulumi a few days, and mainly spend most of my time working in C#

Perhaps it can help someone

requires:
```
"graceful-fs": "^4.2.9",
 "js-yaml": "^4.1.0",
```

Static Config Helper:
```typescript
import * as fs from 'graceful-fs'
import * as yaml from 'js-yaml'

// Some File you have with static config in you wish to share between environments
const staticConfigFilePath = "static-config.yaml";  

export const GetStaticConfig = <T>(key: string): T => {
    const fileContents: any = loadYamlFile(staticConfigFilePath);
    if (fileContents === null || fileContents === undefined) {
        return <T>{};
    }

    if (Object.keys(fileContents).length === 0) {
        return <T>{};
    }

    const indexOfKey = Object.keys(fileContents).indexOf(key);
    if (indexOfKey === null || indexOfKey === undefined) {
        return <T>{};
    }

    const configObject: any = Object.values(fileContents)[indexOfKey];
    if (configObject === null || configObject === undefined) {
        return <T>{};
    }

    return <T>configObject;
};

export const MergeConfigValues = <T>(source: T, target: T): T => {
    if ((source === null || source === undefined) && (target === null || target === undefined)) {
        return <T>{};
    }

    if ((target === null || target === undefined)) {
        return source;
    }

    if ((source === null || source === undefined)) {
        return target;
    }

    return AssignObject(source, target);
}

const parse = (data: any) : any => {
    return yaml.load(data)
}

const loadYamlFile = (path: string): any => {
    return parse(fs.readFileSync(path, 'utf8'))
}

const AssignObject = <T>(target: T, ...sources: T[]): T => {
    sources.forEach(source => {
        Object.keys(source).forEach(key => {
            // @ts-ignore
            const s_val = source[key]
            // @ts-ignore
            const t_val = target[key]
            // @ts-ignore
            target[key] = t_val && s_val && typeof t_val === 'object' && typeof s_val === 'object'
                ? AssignObject(t_val, s_val)
                : s_val
        })
    })
    return target
}
```

Config Provider:
```typescript
import * as pulumi from "@pulumi/pulumi";
import * as staticConfigHelper from './static-config.helper';

export abstract class ConfigurationProvider<TConfig, TSecurityConfig = undefined> {
    private readonly configurationKey: string;
    private readonly securityKey?: string;
    private readonly appName: string;
    private readonly defaultSettings: TConfig;
    
    protected constructor(appName: string, defaultSettings: TConfig, configurationKey: string, securityKey?: string) {
        this.appName = appName;
        this.defaultSettings = defaultSettings;
        this.configurationKey = configurationKey;
        
        if (securityKey) {
            this.securityKey = securityKey;
        }
    }
    
    public GetConfig = (): TConfig => {
        const config: pulumi.Config = new pulumi.Config(this.appName);
        let result = this.defaultSettings;
        const pulumiConfig = config.getObject<TConfig>(this.configurationKey) ?? <TConfig>{};
        const staticConfig = staticConfigHelper.GetStaticConfig<TConfig>(this.configurationKey);

        result = staticConfigHelper.MergeConfigValues<TConfig>(result, pulumiConfig);
        result = staticConfigHelper.MergeConfigValues<TConfig>(result, staticConfig);
        
        return result;
    }
    
    public GetSecurityAsObject = () : pulumi.Output<TSecurityConfig> | undefined => {
        if (this.securityKey === null || this.securityKey === undefined) {
            return undefined;
        }

         // Can't check type of TSecurityConfig here because typescript
        // has no way to check types by type, only by value.
        // This will always be undefined if TSecurityConfig is undefined, but that is expected!
        // In those situations it is expected that if you need to retrieve a security
        // item, you use the string method below, not object.
        
        const config: pulumi.Config = new pulumi.Config(this.appName);
        return config.getSecretObject<TSecurityConfig>(this.securityKey);
    }
    
    public GetSecurityAsString = () : pulumi.Output<string> | undefined => {
        if (this.securityKey === null || this.securityKey === undefined) {
            return undefined;
        }
        
        const config: pulumi.Config = new pulumi.Config(this.appName);
        return config.getSecret(this.securityKey);
    }
}
```

Using the above you can then have a static stack file, which you can "augment" with pulumi stack information.
Secrets is a bit meh - but good for values you wish to have base config for and share between stacks

For instance, you'd build a config provider from the above like so:

```typescript
import {ConfigurationProvider} from "./configurationProvider";

export interface SomeServiceConfig {
    aspnetenvironment: string;
    servicename: string;
}

export class SomeServiceConfigProvider extends ConfigurationProvider<SomeServiceConfig> {
    constructor(appName: string) {
        super(appName, DefaultSettings, 'config');
    }
}

export const DefaultSettings : SomeServiceConfig = {
    aspnetenvironment: 'Development',
    servicename: 'IppexPdfDataMiner',
}
```

**Comment by atrauzzi** [comment]
Disappointed to see that this hasn't been resolved yet.

**Comment by jkodroff** [comment]
Feels like the ESC solves this problem. I would propose we close this issue.
